# **Data Preparation:**

In [69]:
import pandas as pd 
import numpy as np 

In [70]:
data= pd.read_csv("https://raw.githubusercontent.com/rajesh-coventry/Projects/refs/heads/master/00_DataSets/30_Adults_Income_Dataset.csv")

In [71]:
data.shape

(48842, 15)

In [72]:
data.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'gender',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')

In [73]:
data.head(1)

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,salary
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K


| Column             | Meaning                                                                                                                                                                   |
| ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **age**            | Age of the individual (continuous variable).                                                                                                                              |
| **workclass**      | Type of employment or class of work (e.g., 'Private', 'Self-emp', 'Government', etc.). Represents socio-economic status.                                                  |
| **fnlwgt**         | Final weight (used to adjust for survey sampling design). It indicates how many people the census believes a person represents. Often not useful for predictive modeling. |
| **education**      | Highest level of education completed (e.g., 'Bachelors', 'HS-grad', 'Masters', etc.).                                                                                     |
| **education-num**  | Numerical representation of education level (e.g., 'Bachelors' = 13). More machine-readable than `education`.                                                             |
| **marital-status** | Marital status (e.g., 'Married-civ-spouse', 'Divorced', 'Never-married', etc.).                                                                                           |
| **occupation**     | Job type or field (e.g., 'Tech-support', 'Craft-repair', 'Sales', etc.).                                                                                                  |
| **relationship**   | Relationship of the individual to others in the household (e.g., 'Husband', 'Not-in-family', 'Own-child', etc.).                                                          |
| **race**           | Race of the individual (e.g., 'White', 'Black', 'Asian-Pac-Islander', etc.).                                                                                              |
| **gender**         | Gender of the individual ('Male' or 'Female').                                                                                                                            |
| **capital-gain**   | Income from capital gains (e.g., profits from selling stocks or property). Mostly zeros.                                                                                  |
| **capital-loss**   | Losses from capital assets (e.g., losses in investments). Also mostly zeros.                                                                                              |
| **hours-per-week** | Number of hours worked per week (continuous variable).                                                                                                                    |
| **native-country** | Country of origin (e.g., 'United-States', 'Mexico', 'Philippines', etc.).                                                                                                 |
| **salary**         | Target variable: income class. Either '`<=50K`' or '`>50K`', indicating whether the individual earns more or less than \$50,000 annually.                                 |


In [74]:
data.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,salary
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States,>50K
4,18,?,103497,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K


In [ ]:
# Delete fnlwgt column as we won't this in our analysis: 
data.drop(columns=['fnlwgt'], inplace=True)

In [76]:
data.head(3)

,age,workclass,education,education-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,salary
0,25,Private,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K


In [77]:
data.dtypes

age                int64
workclass         object
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
gender            object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
salary            object
dtype: object

In [78]:
# Convert object columns to category
data['workclass'] = data['workclass'].astype('category')

In [79]:
data['education'] = data['education'].astype('category')
data['marital-status'] = data['marital-status'].astype('category')

In [80]:
# generate descriptive statistics for numeric columns: 
data.describe()

,age,education-num,capital-gain,capital-loss,hours-per-week
count,48842.000000,48842.000000,48842.000000,48842.000000,48842.000000
mean,38.643585,10.078089,1079.067626,87.502314,40.422382
std,13.710510,2.570973,7452.019058,403.004552,12.391444
min,17.000000,1.000000,0.000000,0.000000,1.000000
25%,28.000000,9.000000,0.000000,0.000000,40.000000
50%,37.000000,10.000000,0.000000,0.000000,40.000000
75%,48.000000,12.000000,0.000000,0.000000,45.000000
max,90.000000,16.000000,99999.000000,4356.000000,99.000000


**Age:**

1. The total number of observations (rows or data records) in the dataset is **48,842**.

2. The **average (mean) age** of individuals in the dataset is approximately **38.6 years**.

3. The **standard deviation (std)** of **13.7 years** means that **most ages (about 68% under   normal distribution) fall within 13.7 years above or below the mean**.
   So, roughly 68% of people have an age between **(38.6 - 13.7 = 24.9)** and **(38.6 + 13.7 = 52.3)**. It's not a guarantee for a *random person*, but rather a description of the **spread or variability** in the data. If age is not normally distributed, then the interpretation of the standard deviation as a simple `“plus or minus range around the mean”`  does not reliably hold true. In such cases we use `percentiles`, `boxplots`, `histograms` and `density plots` to approximate the distribution of variables (=age in this example). 

4. The **minimum age** recorded is **17 years**.

5. The **maximum age** recorded is **90 years**.

6. The **25th percentile (Q1)** being **28 years** means **25% of the people are aged 28 or younger**.

7. The **50th percentile (median)** being **37 years** means **50% of the people are aged 37 or younger**.

8. The **75th percentile (Q3)** being **48 years** means **75% of the people are aged 48 or younger**.

**Capital Gain:**  

1. The **average capital gain** across all individuals is approximately **1,079**.
   But this is misleading without distribution context (explained below).

2. The **standard deviation** is **7,452**, which indicates **high variability** in capital gains — some people have very high gains while most have none.

3. The **minimum capital gain** is **0** — many people didn’t report any capital gain.

4. The **maximum capital gain** is **99,999**, showing at least one person had a very large gain. This is an extreme outlier.

5. The **25th percentile** is **0**, meaning **25% of the people had zero capital gain**.

6. The **50th percentile (median)** is **0**, so **at least 50% of individuals had no capital gain at all**.

7. The **75th percentile** is also **0**, meaning **at least 75% of the people had zero capital gain**.

Capital Gain is **heavily right-skewed**. Most people had no capital gain, and only a small number had very high values. So:

* The **mean is inflated** by outliers.
* Use **histograms** or **log scale plots** to visualize this skewed distribution.
* Consider analyzing **non-zero gains separately**.

**Capital Loss:**  

1. The **average capital loss** is about **87.5**, but again, that’s not very meaningful due to skewness.

2. The **standard deviation** is **403**, indicating large variability.

3. The **minimum value** is **0**, meaning many individuals had no capital loss.

4. The **maximum capital loss** is **4,356**.

5. The **25th percentile** is **0**, so **25% of the people had zero loss**.

6. The **median** is also **0**, so **at least 50% of individuals had no capital loss**.

7. The **75th percentile** is **0** again, meaning **at least 75% had no capital loss**.

Similar to capital gain, **capital loss is also heavily right-skewed**. Most people didn’t report any losses, and a few had high losses. This again inflates the mean.

In [81]:
# general info of dataset: 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   age             48842 non-null  int64   
 1   workclass       48842 non-null  category
 2   education       48842 non-null  category
 3   education-num   48842 non-null  int64   
 4   marital-status  48842 non-null  category
 5   occupation      48842 non-null  object  
 6   relationship    48842 non-null  object  
 7   race            48842 non-null  object  
 8   gender          48842 non-null  object  
 9   capital-gain    48842 non-null  int64   
 10  capital-loss    48842 non-null  int64   
 11  hours-per-week  48842 non-null  int64   
 12  native-country  48842 non-null  object  
 13  salary          48842 non-null  object  
dtypes: category(3), int64(5), object(6)
memory usage: 4.2+ MB


In [82]:
# check for null values:
data.isnull().sum().any()

np.False_

In [83]:
# check for duplicates:
data.duplicated().sum().any()

np.True_

In [84]:
# count duplicates :
data.duplicated().sum()

np.int64(6374)

In [85]:
# drop duplicates: 
data.drop_duplicates(inplace= True)

In [86]:
data.shape

(42468, 14)

In [87]:
data.head(3)

,age,workclass,education,education-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,salary
0,25,Private,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K


In [88]:
# all the numeric columns:
data.select_dtypes(include=['int64', 'float64']).columns

Index(['age', 'education-num', 'capital-gain', 'capital-loss',
       'hours-per-week'],
      dtype='object')

In [89]:
# Select all the categorical columns: 
data.select_dtypes(include= ["category"]).columns

Index(['workclass', 'education', 'marital-status'], dtype='object')

In [90]:
# Select object type columns:
data.select_dtypes(include= ["object"]).columns

Index(['occupation', 'relationship', 'race', 'gender', 'native-country',
       'salary'],
      dtype='object')

In [91]:
data["workclass"].unique()

['Private', 'Local-gov', '?', 'Self-emp-not-inc', 'Federal-gov', 'State-gov', 'Self-emp-inc', 'Without-pay', 'Never-worked']
Categories (9, object): ['?', 'Federal-gov', 'Local-gov', 'Never-worked', ..., 'Self-emp-inc', 'Self-emp-not-inc', 'State-gov', 'Without-pay']

In [92]:
data["workclass"].value_counts()

workclass
Private             28312
Self-emp-not-inc     3735
Local-gov            3011
?                    2411
State-gov            1927
Self-emp-inc         1644
Federal-gov          1397
Without-pay            21
Never-worked           10
Name: count, dtype: int64

In [93]:
# Change ? workclass to Others:
data["workclass"] = data["workclass"].replace("?", "Others")

C:\Users\MyMachine\AppData\Local\Temp\ipykernel_16068\1899361705.py:2: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.
  data["workclass"] = data["workclass"].replace("?", "Others")


In [94]:
data["workclass"].value_counts()

workclass
Private             28312
Self-emp-not-inc     3735
Local-gov            3011
Others               2411
State-gov            1927
Self-emp-inc         1644
Federal-gov          1397
Without-pay            21
Never-worked           10
Name: count, dtype: int64

In [95]:
data.columns

Index(['age', 'workclass', 'education', 'education-num', 'marital-status',
       'occupation', 'relationship', 'race', 'gender', 'capital-gain',
       'capital-loss', 'hours-per-week', 'native-country', 'salary'],
      dtype='object')

In [96]:
data["education"].nunique()
data["education"].unique()

['11th', 'HS-grad', 'Assoc-acdm', 'Some-college', '10th', ..., 'Assoc-voc', '9th', '12th', '1st-4th', 'Preschool']
Length: 16
Categories (16, object): ['10th', '11th', '12th', '1st-4th', ..., 'Masters', 'Preschool', 'Prof-school', 'Some-college']

In [97]:
data["marital-status"].nunique()
data["marital-status"].unique()

['Never-married', 'Married-civ-spouse', 'Widowed', 'Divorced', 'Separated', 'Married-spouse-absent', 'Married-AF-spouse']
Categories (7, object): ['Divorced', 'Married-AF-spouse', 'Married-civ-spouse', 'Married-spouse-absent', 'Never-married', 'Separated', 'Widowed']

In [98]:
data["occupation"].nunique()
data["occupation"].value_counts()

occupation
Prof-specialty       5679
Exec-managerial      5412
Adm-clerical         4813
Sales                4766
Craft-repair         4726
Other-service        4366
Machine-op-inspct    2492
?                    2421
Transport-moving     2071
Handlers-cleaners    1738
Farming-fishing      1434
Tech-support         1351
Protective-serv       944
Priv-house-serv       240
Armed-Forces           15
Name: count, dtype: int64

In [99]:
# replace ? by others: 
data["occupation"]= data["occupation"].replace("?", "Others")

In [100]:
data["occupation"].value_counts()

occupation
Prof-specialty       5679
Exec-managerial      5412
Adm-clerical         4813
Sales                4766
Craft-repair         4726
Other-service        4366
Machine-op-inspct    2492
Others               2421
Transport-moving     2071
Handlers-cleaners    1738
Farming-fishing      1434
Tech-support         1351
Protective-serv       944
Priv-house-serv       240
Armed-Forces           15
Name: count, dtype: int64

In [101]:
data["relationship"].nunique()
data["relationship"].value_counts()

relationship
Husband           16641
Not-in-family     11384
Own-child          5850
Unmarried          4876
Wife               2242
Other-relative     1475
Name: count, dtype: int64

In [102]:
data["race"].nunique()
data["race"].value_counts()

race
White                 35541
Black                  4536
Asian-Pac-Islander     1515
Amer-Indian-Eskimo      470
Other                   406
Name: count, dtype: int64

In [103]:
data.columns

Index(['age', 'workclass', 'education', 'education-num', 'marital-status',
       'occupation', 'relationship', 'race', 'gender', 'capital-gain',
       'capital-loss', 'hours-per-week', 'native-country', 'salary'],
      dtype='object')

In [104]:
data["gender"].unique()

array(['Male', 'Female'], dtype=object)

In [105]:
data["capital-gain"].nunique()
data["capital-gain"].value_counts()

capital-gain
0        38471
15024      501
7688       400
7298       358
99999      241
         ...  
6612         1
1111         1
22040        1
1639         1
2387         1
Name: count, Length: 123, dtype: int64

Despite being numeric datatype, `capital_gain` also posses the properties of categorical variables.

In [106]:
data["capital-loss"].nunique()

99

In [107]:
data["capital-loss"].value_counts()

capital-loss
0       40203
1902      299
1977      250
1887      229
1485       71
        ...  
1539        1
2489        1
2080        1
155         1
2201        1
Name: count, Length: 99, dtype: int64

In [108]:
data["hours-per-week"].dtype

dtype('int64')

In [109]:
data["native-country"].nunique()

42

In [110]:
data["native-country"].value_counts().index

Index(['United-States', 'Mexico', '?', 'Philippines', 'Germany', 'Puerto-Rico',
       'Canada', 'El-Salvador', 'India', 'Cuba', 'England', 'China', 'South',
       'Jamaica', 'Italy', 'Dominican-Republic', 'Japan', 'Poland', 'Vietnam',
       'Guatemala', 'Columbia', 'Haiti', 'Portugal', 'Taiwan', 'Iran',
       'Nicaragua', 'Greece', 'Peru', 'Ecuador', 'France', 'Ireland',
       'Thailand', 'Hong', 'Cambodia', 'Trinadad&Tobago', 'Laos',
       'Outlying-US(Guam-USVI-etc)', 'Yugoslavia', 'Scotland', 'Honduras',
       'Hungary', 'Holand-Netherlands'],
      dtype='object', name='native-country')

In [111]:
# replace ? by Others:
data["native-country"] = data["native-country"].replace("?", "Others")

In [112]:
data["native-country"].value_counts().index

Index(['United-States', 'Mexico', 'Others', 'Philippines', 'Germany',
       'Puerto-Rico', 'Canada', 'El-Salvador', 'India', 'Cuba', 'England',
       'China', 'South', 'Jamaica', 'Italy', 'Dominican-Republic', 'Japan',
       'Poland', 'Vietnam', 'Guatemala', 'Columbia', 'Haiti', 'Portugal',
       'Taiwan', 'Iran', 'Nicaragua', 'Greece', 'Peru', 'Ecuador', 'France',
       'Ireland', 'Thailand', 'Hong', 'Cambodia', 'Trinadad&Tobago', 'Laos',
       'Outlying-US(Guam-USVI-etc)', 'Yugoslavia', 'Scotland', 'Honduras',
       'Hungary', 'Holand-Netherlands'],
      dtype='object', name='native-country')

In [113]:
data.columns

Index(['age', 'workclass', 'education', 'education-num', 'marital-status',
       'occupation', 'relationship', 'race', 'gender', 'capital-gain',
       'capital-loss', 'hours-per-week', 'native-country', 'salary'],
      dtype='object')

In [114]:
data['salary'].value_counts()

salary
<=50K    32020
>50K     10448
Name: count, dtype: int64

In [115]:
# save the data:
data.to_csv("data.csv", index= False)

------
----
----
-----
-----